### DISCUSSION TITLE
Are there any api/protocol documents?

### DISCUSSION THREAD

**Discussion by lin-calvin** [Xpra-org/xpra#3946] [Q&A] [answered]
I would lile to develop a android client for xpra but there are seems no protocol documents

**Comment by totaam** [comment]
There used to be a couple of documentation pages about the network protocol and packet format on the old wiki.
I didn't convert them when moving to github because they were already out of date.
I do have plans to modify the code to be self-documenting. In particular, the switch to the new packet format (https://github.com/Xpra-org/xpra/issues/1942) should help.

In the meantime, the crucial bits are quite simple:
* make a socket connection
* format all your packets with this header format: https://github.com/Xpra-org/xpra/blob/56cfbee55e6561f57d6d5b61547930da0c2de930/xpra/net/protocol/header.py#L38-L42
* send a `rencodeplus` packet containing `("hello", capabilities)` where `capabilities` is just a dictionary
* you will receive a `hello` packet back using the same format
* handle new packets as they come - ie `new-window`, etc.

There used to be an Android client, but it never really worked well.

**Comment by totaam** [comment, marked as answer]
There is now a work in progress: https://github.com/Xpra-org/xpra/issues/3981
The subsystems documentation is a good start:
https://github.com/Xpra-org/xpra/tree/master/docs/Subsystems
